### PR TITLE
trans: Move rust_try into the compiler

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit moves the IR files in the distribution, rust_try.ll,
rust_try_msvc_64.ll, and rust_try_msvc_32.ll into the compiler from the main
distribution. There's a few reasons for this change:
- LLVM changes its IR syntax from time to time, so it's very difficult to
  have these files build across many LLVM versions simultaneously. We'll likely
  want to retain this ability for quite some time into the future.
- The implementation of these files is closely tied to the compiler and runtime
  itself, so it makes sense to fold it into a location which can do more
  platform-specific checks for various implementation details (such as MSVC 32
  vs 64-bit).
- This removes LLVM as a build-time dependency of the standard library. This may
  end up becoming very useful if we move towards building the standard library
  with Cargo.

In the immediate future, however, this commit should restore compatibility with
LLVM 3.5 and 3.6.

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @brson 

cc @rust-lang/compiler  (my knowledge of trans is pretty bad and I may have done egregious things)

File: src/librustc_trans/trans/intrinsic.rs [hunk 1]
```diff
@@ -989,3 +1026,279 @@ fn with_overflow_intrinsic<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
+        bcx
+    } else if bcx.sess().target.target.options.is_like_msvc {
+        trans_msvc_try(bcx, func, data, dest, dloc)
+    } else {
+        trans_gnu_try(bcx, func, data, dest, dloc)
+    }
+}
+
+// MSVC's definition of the `rust_try` function. The exact implementation here
+// is a little different than the GNU (standard) version below, not ony because
```
Comment (Aatch) [inline src/librustc_trans/trans/intrinsic.rs]: "ony" -> "only"

**Comment by brson** [comment]
Awesome patch.

File: src/librustc_trans/trans/intrinsic.rs [hunk 2]
```diff
@@ -989,3 +1026,304 @@ fn with_overflow_intrinsic<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
+        let ccx = bcx.ccx();
+        let dloc = DebugLoc::None;
+
+        // Type indicator for the exception being thrown, not entirely sure
+        // what's going on here but it's what all the examples in LLVM use.
+        let lpad_ty = Type::struct_(ccx, &[Type::i8p(ccx), Type::i32(ccx)],
+                                    false);
+
+        // Define the "inner try" shim
+        let rust_try_inner = declare::declare_internal_rust_fn(ccx,
```
Comment (nagisa) [inline src/librustc_trans/trans/intrinsic.rs]: If you are “defining” the shim, `declare::define_internal_rust_fn` should be used.

File: src/librustc_trans/trans/intrinsic.rs [hunk 3]
```diff
@@ -989,3 +1026,304 @@ fn with_overflow_intrinsic<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
+        trans_rust_try(ccx, rust_try_inner, lpad_ty, bcx.fcx.eh_personality(),
+                       output, dloc, &mut |bcx, then, catch| {
+            let func = llvm::get_param(rust_try_inner, 0);
+            let data = llvm::get_param(rust_try_inner, 1);
+            Invoke(bcx, func, &[data], then.llbb, catch.llbb, None, dloc);
+            C_null(Type::i8p(ccx))
+        });
+
+        // Define the "outer try" shim.
+        let rust_try = declare::declare_internal_rust_fn(ccx, "__rust_try",
```
Comment (nagisa) [inline src/librustc_trans/trans/intrinsic.rs]: ditto.

File: src/rustllvm/RustWrapper.cpp [hunk 4]
```diff
@@ -120,7 +120,8 @@ extern "C" void LLVMAddDereferenceableCallSiteAttr(LLVMValueRef Instr, unsigned
                                                          idx, B)));
 }
 
-extern "C" void LLVMAddFunctionAttribute(LLVMValueRef Fn, unsigned index, uint64_t Val) {
+extern "C" void LLVMAddFunctionAttribute(LLVMValueRef Fn, unsigned index,
```
Comment (nagisa) [inline src/rustllvm/RustWrapper.cpp]: nit: doesn’t seem to be necessary change.

**Comment by vadimcn** [comment]
:(
Couldn't we've ran .ll files through cpp (like is done with regular asm) for the same effect?

**Comment by alexcrichton** [comment]
In theory, yes, but that still requires the standard library to have `llc` available to compile itself, which we need to move away from if we want to build std with Cargo at some point. It also would have been interesting trying to detect the llc version, pass the right variables to the preprocessor, and then piping all that back into llc to create an object.

**Comment by vadimcn** [comment]
> In theory, yes, but that still requires the standard library to have llc available to compile itself

I am not sure I understand this point.  Didn't `rust_try`live in `rustrt_native` library?  Which is effectively a part of the compiler, which already requires LLVM to be around. 
How it it different from all other runtime fns in `rustrt_native` and `compiler-rt`?

My concern with the new way is that:
- Emitting IR via LLVM API takes a lot more work and is way less readable because of all the noise added by API usage.
- One cannot override `rust_try` for a custom target, like you we able to do for `*-pc-windows-msvc) without changing the compiler.

**Comment by vadimcn** [comment]
(sorry for jumping in late - I was not aware of this change until bors told me that my PR collided with this one)

**Comment by alexcrichton** [comment]
> I am not sure I understand this point. Didn't rust_trylive in rustrt_native library? Which is effectively a part of the compiler, which already requires LLVM to be around.
How it it different from all other runtime fns in rustrt_native and compiler-rt?

Ah by this I just mean that the tool `llc` is no longer needed to build the standard library.

The other shims and such in `rustrt_native` can be built with a normal C compiler and I believe that all the shims in `compiler-rt` can also be built with a normal C compiler. If you're in theory cross compiling the standard library to a new architecture it's much more likely you've got the right C compiler installed vs `llc`.

> Emitting IR via LLVM API takes a lot more work and is way less readable because of all the noise added by API usage.

I don't disagree with this!

> One cannot override rust_try for a custom target, (like you we able to do for *-pc-windows-msvc) without changing the compiler.

I'm not quite sure I follow this. Is this basically saying that it's easier to edit IR than it is to edit the compiler?

> sorry for jumping in late - I was not aware of this change until bors told me that my PR collided with this one

No worries!

**Comment by vadimcn** [comment]
> I'm not quite sure I follow this. Is this basically saying that it's easier to edit IR than it is to edit the compiler?

I meant that as a library function in rustrt_native, it could have been overridden.  Not so much with it as compiler intrinsic.
In principle, rust_try could have been implemented in C++, assembly, or even in C, if the compiler supports suitable extensions.  It's just hard to do so portably.

**Comment by alexcrichton** [comment]
How do you override something in a native library? You mean avoiding `#[cfg(stage0)]`?

The ABI of unwinding is very closely tied to the compiler itself, so even though it could be implemented in various locations it kinda makes sense to me to implement this as an intrinsic so the compiler has complete control over all ABI possibilities.

**Comment by vadimcn** [comment]
> How do you override something in a native library? 

I meant that when building rustrt_native for a new target, you could write a custom rust_try.

**Comment by alexcrichton** [comment]
That's true, but do you know if that's done often? From what I've seen the `rust_try.ll` file basically never changes.

**Comment by eddyb** [comment]
Was it known that this would break unwinding entirely on `-Z no-landing-pads` builds?

**Comment by eddyb** [comment]
Actually, scratch that, it's because the snapshot compiler doesn't have the `try` intrinsic so this PR uses a noop, breaking unwinding for the `stage1` compiler and tests.

**Comment by alexcrichton** [comment]
@eddyb yes this was known.

**Pull request merged**